File: disc-math/lab1/A.cpp
#include <iostream>
#include <unordered_set>
#include <vector>

using namespace std;

void print_int_vector(const vector<int> & vector) {
    for (int el : vector) {
        cout << el + 1 << " ";
    }
}

bool exist_edge(const vector<unordered_set<int>> & graph, int i, int j) {
    if (i > j) {
        return graph[i].find(j) != graph[i].end();
    } else {
        return graph[j].find(i) != graph[j].end();
    }
}

void solve() {
    int n;
    cin >> n;
    vector<unordered_set<int>> graph(n);
    vector<int> queue(n);

    for (int i = 0; i < n; ++i) {
        queue[i] = i;

        if (i == 0) {
            continue;
        }

        string vertexes;
        cin >> vertexes;
        for (int j = 0; j < vertexes.size(); ++j) {
            if (vertexes[j] == '1') {
                graph[i].insert(j);
            }
        }
    }

    for (int k = 0; k < n * (n - 1); ++k) {
        if (!exist_edge(graph, queue[0], queue[1])) {
            int i = 2;
            while (!exist_edge(graph, queue[0], queue[i]) || !exist_edge(graph, queue[1], queue[i + 1])) {
                ++i;
            }

            for (int j = 0; j < i / 2; ++j) {
                int tmp = queue[1 + j];
                queue[1 + j] = queue[i - j];
                queue[i - j] = tmp;
            }
        }

        queue.push_back(queue[0]);
        queue.erase(queue.begin());
    }

    print_int_vector(queue);
}

int main() {
    solve();
}

File: disc-math/lab1/A.py
def exist_edge(graph, i, j):
    if i > j:
        return j in graph[i]
    else:
        return i in graph[j]

def solve():
    n = int(input())
    graph = [set() for i in range(n)]
    queue = [i for i in range(n)]

    for i in range(n):
        vertex = input()
        for k in range(len(vertex)):
            if vertex[k] == '1':
                graph[i].add(k)

    for k in range(4 * n):
        if not exist_edge(graph, queue[0], queue[1]):
            i = 2
            while not exist_edge(graph, queue[0], queue[i]) or not exist_edge(graph, queue[1], queue[i + 1]):
                i += 1
            queue[1:i + 1] = queue[1: i + 1][::-1]
        queue.append(queue.pop(0))

    print(' '.join([str(i + 1) for i in queue]))

solve()
